protocols/gossipsub/src/peer_score/params.rs

```diff
@@ -0,0 +1,399 @@
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+use crate::TopicHash;
+use std::collections::{HashMap, HashSet};
+use std::net::IpAddr;
+use std::time::Duration;
+
+/// The default number of seconds for a decay interval.
+const DEFAULT_DECAY_INTERVAL: u64 = 1;
+/// The default rate to decay to 0.
+const DEFAULT_DECAY_TO_ZERO: f64 = 0.1;
+
+/// Computes the decay factor for a parameter, assuming the `decay_interval` is 1s
+/// and that the value decays to zero if it drops below 0.01.
+pub fn score_parameter_decay(decay: Duration) -> f64 {
+    score_parameter_decay_with_base(
+        decay,
+        Duration::from_secs(DEFAULT_DECAY_INTERVAL),
+        DEFAULT_DECAY_TO_ZERO,
+    )
+}
+
+/// Computes the decay factor for a parameter using base as the `decay_interval`.
+pub fn score_parameter_decay_with_base(decay: Duration, base: Duration, decay_to_zero: f64) -> f64 {
+    // the decay is linear, so after n ticks the value is factor^n
+    // so factor^n = decay_to_zero => factor = decay_to_zero^(1/n)
+    let ticks = decay.as_secs_f64() / base.as_secs_f64();
+    decay_to_zero.powf(1f64 / ticks)
+}
+
+#[derive(Debug, Clone)]
+pub struct PeerScoreThresholds {
+    /// The score threshold below which gossip propagation is suppressed;
+    /// should be negative.
+    pub gossip_threshold: f64,
+
+    /// The score threshold below which we shouldn't publish when using flood
+    /// publishing (also applies to fanout peers); should be negative and <= `gossip_threshold`.
+    pub publish_threshold: f64,
+
+    /// The score threshold below which message processing is suppressed altogether,
+    /// implementing an effective graylist according to peer score; should be negative and
+    /// <= `publish_threshold`.
+    pub graylist_threshold: f64,
+
+    /// The score threshold below which px will be ignored; this should be positive
+    /// and limited to scores attainable by bootstrappers and other trusted nodes.
+    pub accept_px_threshold: f64,
+
+    /// The median mesh score threshold before triggering opportunistic
+    /// grafting; this should have a small positive value.
+    pub opportunistic_graft_threshold: f64,
+}
+
+impl Default for PeerScoreThresholds {
+    fn default() -> Self {
+        PeerScoreThresholds {
+            gossip_threshold: -10.0,
+            publish_threshold: -50.0,
+            graylist_threshold: -80.0,
+            accept_px_threshold: 10.0,
+            opportunistic_graft_threshold: 20.0,
```

Comment:
@AgeManning could you please share how those values were defined?

Reply:
These essentially can be defined for whatever network you are running gossipsub on. Its been a few years, but I believe these were from the go-implementation at the time, which may have been used in some of the tests in the v1.1 gossipsub paper where they tested a few attack vectors. 

These numbers shouldn't be set independently, but based on the maximum and minimum values you can get from the score.

Reply:
I found this PR that says the values were inspired by LH https://github.com/status-im/nimbus-eth2/pull/2091. These are the values used:

```
     gossipThreshold: -4000,
     publishThreshold: -8000,
     graylistThreshold: -16000
```

They are used here indeed https://gist.github.com/blacktemplar/5c1862cb3f0e32a1a7fb0b25e79e6e2c#file-51000-toml-L33, but LH doesn't use to configure GossipSub, but rather here https://github.com/libp2p/rust-libp2p/blob/25d14349cc32f7a8e3feffb931ba2e84cccb0013/protocols/gossipsub/src/metrics.rs#L76

Reply:
Now I"m confused about what you are asking. 

The values in this repository i.e `gossip_threshold: -10` is not the same value as what we have chosen in the Ethereum consensus network i.e `-4000`. 

As I mentioned, these values need to be defined for whatever network you are running. For Ethereum's consensus network, we did a bunch of estimations and determined a bunch of numbers. That is entirely independent of the `gossipsub` protocol, which should run on any network. 

If you question is specific to lighthouse or the Ethereum network, then that is a different question and probably should be asked on the lighthouse repo as those values are for a specific network. 

You did point out that the metrics value uses the Ethereum consensus values as default. This probably should be using the same defaults as the rust-libp2p repo. At the time, Lighthouse was the only one consuming these metrics and so was built around our scoring values to get sane results in the histograms.

Reply:
You're right, sorry for the confusion. It's clear now, thanks. I somehow misled myself looking at this PR that introduces params in the lighthouse rust-libp2p fork https://github.com/sigp/rust-libp2p/pull/556/files#diff-9f622a0085d0d96aef9e883b7afdcdd4a652e8801be7cfabf82819f5a74a70e0. Is there a lighthouse PR defining the values for lighthouse?

Reply:
What specifically are you interested in?

Are you asking how we figured out what values to use in Lighthouse? There is an external document I can share with you if that's the case. The PR to lighthouse will likely just show values being added with no context. 

Feel free to DM me in the Lighthouse discord if you like